[우] API 테스트/문서자동화 미션 제출합니다.

hwookim opened the pull request:
제공된 코드를 기반으로 하다 보니 컨벤션이 일부 꼬인 점이 있습니다.ㅠㅠ
ATDD 방식을 적용해 인수테스트를 작성하고 out-in 방식으로 각 단위 테스트를 진행하였습니다.
각 단위 테스트를 진행하긴 했는데 사이드 케이스에 대한 테스트가 좀 부실한 것 같네요.
리팩토링하며 추가하도록 하겠습니다.

리뷰 잘 부탁드립니다!

----
리팩토링 도중 궁금한 게 생겨 질문 남깁니다.

즐겨찾기 기능에 대한 테스트를 진행할 때, 예외 사항을 어느 정도까지 커버해야 할지 궁금합니다.
현재 즐겨찾기 추가 기능은 경로 검색을 한 이후에 실행할 수 있는데요.
덕분에 갈 수 없는 경로, 출발지와 도착지가 같거나 null이 들어간 경우 등에 대하여 경로 검색 부분에서 검증이 모두 되어있습니다.
전체적으로 볼 때는 경로 검색 / 즐겨찾기 추가를 한 로직으로 볼 수 있을 것 같습니다.
하지만 결국 별도의 API기 때문에 각기 검증 절차를 거쳐야 할까요?
이는 경로 검색에 대한 로직을 실제 경로 검색 시, 즐겨찾기 추가 시, 총 두 번 일어나게 합니다.
그렇다고 추가하지 않자니, 프론트 설계가 바뀌어 즐겨찾기 추가 방식이 변경된다면 이슈가 발생할 수 있을 것 같네요.
예외 테스트의 범위를 어느 정도까지 가져가야 할까요?

wmakerjun committed:
feat(index): change index page endpoint from admin to service

wmakerjun committed:
refactor(mypage): change my page endpoint from '/me' to '/mypage'

wmakerjun committed:
feat(join): add join page with ajax

wmakerjun committed:
feat(login): add login page with ajax

wmakerjun committed:
feat(mypage): add mypage with ajax

wmakerjun committed:
feat(api): update service api

wmakerjun committed:
feat(favorite): update favorite page for read and delete

wmakerjun committed:
refactor(servie): update promise into async/await

moonyoungCHAE committed:
refactor: admin page 변경

moonyoungCHAE committed:
feat: Bearer 인증 기능 추가 및 테스트, 깨지는 테스트 수정

moonyoungCHAE committed:
test: Member Controller create 테스트 및 문서화

moonyoungCHAE committed:
docs: 1단계 기능 구현 목록 작성

moonyoungCHAE committed:
test: Member findByEmail 테스트

moonyoungCHAE committed:
test: Member Update 테스트

moonyoungCHAE committed:
test: Member Delete 테스트

moonyoungCHAE committed:
test: MemberControllerTest 테스트 환경 변경

moonyoungCHAE committed:
test: LoginMemberController 에서 Token 테스트

leewnsdud committed:
refactor: MemberControllerTest 테스트 검증 방식 수정

moonyoungCHAE committed:
refactor: package 이동

moonyoungCHAE committed:
feat: MemberController 분리 및 인수테스트 작성

 - LoginMemberController / AdminMemberController로 분리
 - LoginMemberController 인수테스트 작성 및 해당 기능 구현

moonyoungCHAE committed:
docs: to do list 작성 (Login 멤버의 회원 관리)

moonyoungCHAE committed:
test: LoginMemberControllerTest 구현

moonyoungCHAE committed:
docs: to-do list 체크 및 추가

moonyoungCHAE committed:
refactor: Interceptor에 대한 단위테스트 작성

- Request에 token이 존재하지 않는 경우에 대한 케이스 테스트
- Request에 비정상적인 token이 존재하는 경우에 대한 케이스 테스트

moonyoungCHAE committed:
refactor: Interceptor에 대한 단위테스트 작성

- Request에 token이 존재하지 않는 경우에 대한 케이스 테스트
- Request에 비정상적인 token이 존재하는 경우에 대한 케이스 테스트

moonyoungCHAE committed:
feat: AdminMemberControllerTest 문서화

 - api-guide.adoc 수정

moonyoungCHAE committed:
feat: LoginMemberControllerTest 문서화

 - 패키지명 변경 doc -> docs

moonyoungCHAE committed:
refactor: PageController, MemberPageController 통합

moonyoungCHAE committed:
docs: front 연동 세부사항 추가

moonyoungCHAE committed:
feat: 로그인, 회원정보 수정, 탈퇴 - Front 연동

moonyoungCHAE committed:
docs: 추가 기능구현목록 작성

moonyoungCHAE committed:
docs: 2단계 기능 구현 목록 작성

moonyoungCHAE committed:
refactor: 메서드 이동 및 잘못된 URI 수정

moonyoungCHAE committed:
test: 즐겨찾기 관리에 대한 인수테스트 작성 (mock controller)

moonyoungCHAE committed:
refactor: MockMvc 한글 인코딩 설정 및 프린트 중복 제거

moonyoungCHAE committed:
feat: FavoriteController 구현 및 테스트

moonyoungCHAE committed:
feat: FavoriteService 구현 및 테스트

 패키지 이동 favorite -> member/favorite

moonyoungCHAE committed:
feat: 즐겨찾기 관리 기능 구현, DB 추가, 테스트

moonyoungCHAE committed:
feat: 즐겨찾기 API Front 연동

moonyoungCHAE committed:
feat: 즐겨찾기 API 문서화

hwookim committed:
docs: To-do List 수정

hwookim committed:
refactor: 패키지 이동

hwookim committed:
refactor: 불필요 메소드 삭제

hwookim committed:
refactor: gson 상수화

hwookim committed:
refactor: 불필요한 개행 제거

kingbbode commented on build.gradle:
```diff
@@ -19,13 +20,33 @@ dependencies {
 	implementation 'pl.allegro.tech.boot:handlebars-spring-boot-starter:0.3.0'
 	implementation 'org.jgrapht:jgrapht-core:1.0.1'
 	implementation 'io.jsonwebtoken:jjwt:0.9.1'
+	implementation 'com.google.code.gson:gson:2.8.6'
 	testImplementation 'io.rest-assured:rest-assured:3.3.0'
 	testImplementation('org.springframework.boot:spring-boot-starter-test') {
 		exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
 	}
 	runtimeOnly 'com.h2database:h2'
+	asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.4.RELEASE'
+	testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:2.0.4.RELEASE'
```
의존성의 버전을 직접 명시해주었네요.
스프링부트에서 관리해주고 있는 버전관리에 대한 부분을 찾아보시면 좋을 듯 합니다 .

hwookim replied:
스프링부트에서 관리하는 의존성에 대해서는 버전을 명시하지 않도록 변경했습니다.

kingbbode commented on src/main/java/wooteco/subway/domain/member/Favorite.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Objects;
+
+public class Favorite {
+    private String source;
+    private String target;
```
source 와 target 의 역이름을 그대로 저장하고 있네요.

신천이란 역을 만들고 즐겨찾기를 추가하였다고 할 때, 신천역 이름이 잠실새내역으로 변경이 되면 사용자의 즐겨찾기는 어떻게 될까요?
데이터 저장소에 저장되는 모든 데이터는 데이터의 키를 중심으로 관리되어야 합니다 :)

Favorite 이 Station 과 연관관계를 맺도록 재설계 부탁드립니다.

hwookim replied:
ID를 통해 즐겨찾기를 관리하도록 변경했습니다.

처음에는 즐겨찾기 검색 시 해당 ID의 지하철 역이 없으면 예외를 반환하도록 했는데, 즐겨찾기 등록 후 지하철 역이 삭제되는 경우도 상정하여 "삭제된 역"이라는 문구를 반환하도록 변경했습니다.

kingbbode commented:
중복에 대한 요구사항은 없지만, 사용자에게 더 좋은 사용성을 제공하기 위해  중복된 즐겨찾기는 등록 안되도록 해보는 것이 어떨까요?

kingbbode reviewed (changes requested):
안녕하세요 :)
코멘트 작성해두었습니다.
확인 부탁드려요!

kingbbode commented:
테스트는 테스트 대상에 대하여 독립적으로 작성되어야 합니다 :)

`프론트 설계가 바뀌어 즐겨찾기 추가 방식이 변경된다면 이슈가 발생할 수 있을 것 같네요.`

이미 잘 이해하고 계신 것 같네요!

hwookim committed:
refactor: 의존성 관리

 미사용 의존성 제거
 버전 관리 위임

hwookim committed:
fix: 즐겨찾기 추가 시 name이 아닌 id를 보내던 문제 해결

hwookim committed:
refactor: Favorite가 StationId를 갖도록 변경

 source, target -> sourceId/Name, targetId/Name 명시
 create 시 request가 id로 바뀌어 stationRepository를 통해 name 구하는 로직 추가
 FavoriteDeleteRequest 추가 구
 데이터 스키마 및 프론트 요청 변경
 테스트 구조 변경

hwookim committed:
feat: 중복되는 즐겨찾기 등록에 대한 검출 로직 추가

hwookim committed:
feat: Advice 추가

hwookim committed:
refactor: 지하철 역이 삭제된 경우에 대비

 FavoriteService.findFavorite() 시 response로 "삭제된 역"을 표시 후 해당 favorite 삭제
 FavoriteService.deleteFavorite() 시 이미 삭제된 역일 경우 프론트 반영 후 로직 종료

hwookim committed:
refactor: 즐겨찾기 등록 시 가능한 경로인지 검

 FavoriteService.pathService 추가 및 검증 이용
 FavoriteServiceTest에 line과 lineStation 추가

kingbbode reviewed (approved):
고생 많으셨습니다! 이만 머지할게요 :)
